server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java

```diff
@@ -115,4 +127,18 @@ public void onFailure(Exception exception) {
             );
         }
     }
+
+    private void addKnnResultsToQuery(SearchSourceBuilder source) {
+        if (source == null || source.knnSearch() == null) {
+            return;
+        }
+
```

Comment:
May be we can create an appropriate QueryBuilder once for all shards and then pass to each ShardSearchRequest?